File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js
function solution(friends, gifts) {
    const n = friends.length;
    var fobj = {};
    friends.forEach((f, idx) => fobj[f] = idx);
    var giftcnt = Array.from(Array(n), () => Array(n).fill(0)); //i가 j에게 준선물수
    var giftPoints = Array(n).fill(0); //선물지수
    var points = Array(n).fill(0); //다음달받을선물
    
    for (const g of gifts){
        const giver = g.split(" ")[0], recver = g.split(" ")[1];
        const i = fobj[giver], j = fobj[recver];
        giftcnt[i][j] += 1;
        giftPoints[i] += 1;
        giftPoints[j] -= 1;
    }
    
    for (let i = 0; i < n; i++){
        for (let j = i + 1; j < n; j++){
            if (giftcnt[i][j] > giftcnt[j][i]) points[i]++; //i가 선물을 더 많이 줬는가
            else if (giftcnt[j][i] > giftcnt[i][j]) points[j]++; //j가 선물을 더 많이 줬는가
            else if (giftPoints[i] > giftPoints[j]) points[i]++; //i가 선물지수가 큰가
            else if (giftPoints[j] > giftPoints[i]) points[j]++; //j가 선물지수가 큰가   
        }
    }
    
    return Math.max(...points);
    
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        final int fNum = friends.length;
        
        Map<String, Integer> findex = new HashMap<>();
        for (int i = 0; i < friends.length; i++) {
            findex.put(friends[i], i);
        }
        
        int[][] count = new int[fNum][fNum];
        int[] give = new int[fNum];
        int[] take = new int[fNum];
        
        for (String g : gifts) {
            String[] sp = g.split(" ");
            int giver = findex.get(sp[0]);
            int taker = findex.get(sp[1]);
            count[giver][taker]++;
            give[giver]++;
            take[taker]++;
        }
        
        int[] nextMonth = new int[fNum];
        for (int i = 0; i < fNum; i++) {
            for (int j = i + 1; j < fNum; j++) {
                if (count[i][j] > count[j][i]) {
                    nextMonth[i]++;
                } else if (count[i][j] < count[j][i]) {
                    nextMonth[j]++;
                } else {
                    int icount = give[i] - take[i];
                    int jcount = give[j] - take[j];
                    if (icount > jcount) {
                        nextMonth[i]++;
                    } else if (jcount > icount) {
                        nextMonth[j]++;
                    }
                }
            }
        }
        
        int answer = 0;
        for (int v : nextMonth) {
            if (v > answer) {
                answer = v;
            }
        }
        
        return answer;
    }
}